src/app/core/components/redeem-gift/redeem-gift.component.ts

```diff
@@ -0,0 +1,91 @@
+/* @format */
+import { Component, Input, OnInit } from '@angular/core';
+import {
+  UntypedFormGroup,
+  UntypedFormBuilder,
+  Validators,
+} from '@angular/forms';
+import { PromoVOData } from '@models';
+import { ApiService } from '@shared/services/api/api.service';
+import {
+  BillingResponse,
+  AccountResponse,
+} from '@shared/services/api/index.repo';
+import { FileSizePipe } from '@shared/pipes/filesize.pipe';
+import { AccountService } from '@shared/services/account/account.service';
+import { EventService } from '@shared/services/event/event.service';
+
+interface StorageRedemptionMessage {
+  successful: boolean;
+  message: string;
+}
+
+@Component({
+  selector: 'pr-redeem-gift',
+  templateUrl: './redeem-gift.component.html',
+  styleUrl: './redeem-gift.component.scss',
+})
+export class RedeemGiftComponent implements OnInit {
+  @Input() public promoCode: string = '';
+  public resultMessage: StorageRedemptionMessage | undefined;
+  public promoForm: UntypedFormGroup;
+  public waiting: boolean;
+
+  constructor(
+    private fb: UntypedFormBuilder,
+    private account: AccountService,
+    private api: ApiService,
+    private event: EventService,
+  ) {}
+
+  public ngOnInit(): void {
+    this.promoForm = this.fb.group({
+      code: [this.promoCode, [Validators.required]],
+    });
+  }
+
+  public async onPromoFormSubmit(value: PromoVOData) {
+    try {
+      this.waiting = true;
+      const response = await this.api.billing.redeemPromoCode(value);
+      await this.handleValidPromoCode(response);
+    } catch (err: unknown) {
+      this.showPromoCodeError(err);
+    } finally {
+      this.waiting = false;
+    }
+  }
+
+  private showPromoCodeError(err: unknown) {
+    const message =
+      err instanceof BillingResponse || err instanceof AccountResponse
+        ? err.getMessage()
+        : 'There was an error redeeming your code.';
+    this.resultMessage = { message, successful: false };
+  }
+
+  private async handleValidPromoCode(response: BillingResponse) {
+    await this.updateAccountStorageBytes(response);
+    this.event.dispatch({
+      entity: 'account',
+      action: 'submit_promo',
+    });
+    this.promoForm.reset();
+  }
+
+  private async updateAccountStorageBytes(response: BillingResponse) {
+    await this.account.refreshAccount();
+    const promo = response.getPromoVO();
+    const bytes = promo.sizeInMB * (1024 * 1024);
+    this.account.addStorageBytes(bytes);
+    this.showPromoCodeSuccess(bytes);
```

Comment:
It feels like this call is more part of "handling a valid promo code" than "updating account storage"; should it be in `handleValidPromoCode` instead?

Reply:
The main reason it's called here is that we want to show the user how much storage space was redeemed. I didn't want to make the redeemed bytes a property of the class and it also felt weird to return the bytes from this method as well.